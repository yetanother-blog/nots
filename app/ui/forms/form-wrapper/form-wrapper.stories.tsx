import type { Meta, StoryObj } from '@storybook/react';
import { FormWrapper } from './form-wrapper';
import { Label } from '../label/label';
import { FormFieldStyle } from '../form-field-style/form-field-style';

/**
 * The FormWrapper component is used to provide a consistent positioning for the
 * form elements.
 */
const meta: Meta<typeof FormWrapper> = {
  title: 'Forms/FormWrapper',
  component: FormWrapper,
};

export default meta;
type Story = StoryObj<typeof FormWrapper>;

/**
 * In this example we compose everything together by adding a `Label`, a
 * `FormFieldStyle` and an `input` element as `children` to the `FormWrapper`
 * component.
 */
export const Primary: Story = {
  render: () => (
    <FormWrapper helperText="Helper text">
      <Label htmlFor="name">Name</Label>
      <FormFieldStyle className="h-11">
        <input
          className="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
          id="name"
          type="text"
        />
      </FormFieldStyle>
    </FormWrapper>
  ),
};

/**
 * Provide an `isError` prop to the `FormWrapper` component to display the error
 * state by using the `helpText` prop.
 */
export const ErrorState: Story = {
  render: () => (
    <FormWrapper helperText="Helper text" isError={true}>
      <Label htmlFor="name">Name</Label>
      <FormFieldStyle className="h-11" isError={true}>
        <input
          className="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
          id="name"
          type="text"
        />
      </FormFieldStyle>
    </FormWrapper>
  ),
};
