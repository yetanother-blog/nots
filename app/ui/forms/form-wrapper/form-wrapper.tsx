import type { FC, PropsWithChildren, ReactNode } from 'react';
import { Typography } from '~/ui';
import { clsx } from 'clsx';

export type FormWrapperProps = {
  helperText?: ReactNode;
  isError?: boolean;
  className?: string;
} & PropsWithChildren;

export const FormWrapper: FC<FormWrapperProps> = ({
  children,
  helperText,
  isError,
  className,
}) => {
  return (
    <div className={clsx('flex flex-col gap-2', className)}>
      {children}
      {helperText && (
        <Typography
          variant="bodyTiny"
          textColor={isError ? 'danger-500' : 'grey-600'}
        >
          {helperText}
        </Typography>
      )}
    </div>
  );
};
