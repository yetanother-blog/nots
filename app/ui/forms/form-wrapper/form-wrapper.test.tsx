import { render } from '@testing-library/react';
import { FormWrapper } from './form-wrapper';

describe('FormWrapper', () => {
  it('should render form wrapper', () => {
    const { container } = render(<FormWrapper>FormWrapper</FormWrapper>);
    expect(container).toMatchSnapshot();
  });

  it('should render form wrapper with helper text', () => {
    const { container } = render(
      <FormWrapper helperText="Helper text">FormWrapper</FormWrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render form wrapper with error', () => {
    const { container } = render(
      <FormWrapper isError={true} helperText="Helper text">
        FormWrapper
      </FormWrapper>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render with custom class name', () => {
    const { container } = render(
      <FormWrapper className="custom-class-name">FormWrapper</FormWrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
