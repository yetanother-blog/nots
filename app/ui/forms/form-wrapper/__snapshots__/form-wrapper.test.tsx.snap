// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormWrapper > should render form wrapper 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    FormWrapper
  </div>
</div>
`;

exports[`FormWrapper > should render form wrapper with error 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    FormWrapper
    <p
      class="antialiased font-sans text-xs font-normal text-nots-danger-500"
    >
      Helper text
    </p>
  </div>
</div>
`;

exports[`FormWrapper > should render form wrapper with helper text 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    FormWrapper
    <p
      class="antialiased font-sans text-xs font-normal text-nots-grey-600"
    >
      Helper text
    </p>
  </div>
</div>
`;

exports[`FormWrapper > should render with custom class name 1`] = `
<div>
  <div
    class="flex flex-col gap-2 custom-class-name"
  >
    FormWrapper
  </div>
</div>
`;
